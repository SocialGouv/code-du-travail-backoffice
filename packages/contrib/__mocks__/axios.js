import "axios";

jest.mock("axios", () => {
  const axios = {
    interceptors: {
      response: {
        use: () => void 0
      }
    },

    delete: async () => void 0,
    get: async () => void 0,
    patch: async () => void 0,
    post: async () => void 0,
    put: async () => void 0
  };

  jest.spyOn(axios, "delete");
  jest.spyOn(axios, "get");
  jest.spyOn(axios, "patch");
  jest.spyOn(axios, "post");
  jest.spyOn(axios, "put");

  return {
    ...axios,
    create: () => axios
  };
});
