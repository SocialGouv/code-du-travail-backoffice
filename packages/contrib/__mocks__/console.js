/* eslint-disable no-undef */

global.console = {
  debug: jest.fn(),
  error: jest.fn(),
  info: jest.fn(),
  log: jest.fn(),
  warn: jest.fn()
};
