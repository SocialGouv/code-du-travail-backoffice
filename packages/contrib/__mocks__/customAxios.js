import "../src/libs/customAxios";

jest.mock("../src/libs/customAxios", () => () => {
  const customAxios = {
    delete: async () => void 0,
    get: async () => void 0,
    patch: async () => void 0,
    post: async () => void 0,
    put: async () => void 0
  };

  jest.spyOn(customAxios, "delete");
  jest.spyOn(customAxios, "get");
  jest.spyOn(customAxios, "patch");
  jest.spyOn(customAxios, "post");
  jest.spyOn(customAxios, "put");

  return customAxios;
});
