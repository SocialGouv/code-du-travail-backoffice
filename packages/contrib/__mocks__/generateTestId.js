global.generateTestId = require("uuid/v4");
