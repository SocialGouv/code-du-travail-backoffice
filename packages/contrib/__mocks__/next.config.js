module.exports = {
  serverRuntimeConfig: {},
  publicRuntimeConfig: {
    API_URI: "http://localhost:3200"
  }
};
