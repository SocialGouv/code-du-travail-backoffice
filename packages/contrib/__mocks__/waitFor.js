global.waitFor = delayInMs =>
  new Promise(resolve => setTimeout(resolve, delayInMs));
