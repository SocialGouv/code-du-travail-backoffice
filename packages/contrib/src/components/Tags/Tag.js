import styled from "@emotion/styled";
import React from "react";
import { Flex } from "rebass";

const Container = styled(Flex)`
  background-color: lightgray;
  border: solid 1px gray;
  border-radius: 0.25rem;
  font-size: 0.875rem;
  margin: 0.5rem 0.5rem 0 0;
  max-width: 32rem;
  padding: 0.2rem 0.4rem;
  user-select: none;
`;
const Text = styled.span`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;
const Button = styled.img`
  cursor: pointer;
  height: 0.75rem;
  margin-left: 0.5rem;
  opacity: 0.5;
  vertical-align: 2px;
  width: 0.75rem;

  :hover {
    opacity: 1;
  }
`;

export default class Tag extends React.PureComponent {
  openUrl(url) {
    window.open(url, "_blank");
  }

  render() {
    const { onRemove, url, value } = this.props;
    const ariaName = this.props.ariaName !== undefined ? this.props.ariaName : "l'Ã©tiquette";

    return (
      <Container alignItems="center">
        <Text>{value}</Text>
        {typeof url === "string" && (
          <Button
            alt={`Bouton ouvrant l'url de ${ariaName} ${value}`}
            onClick={() => this.openUrl(url)}
            src="/static/images/link.svg"
          />
        )}
        <Button
          alt={`Bouton supprimant ${ariaName} ${value}`}
          onClick={() => onRemove(value)}
          src="/static/images/delete.svg"
        />
      </Container>
    );
  }
}
