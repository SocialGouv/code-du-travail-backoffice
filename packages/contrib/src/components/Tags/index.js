import { any, prop, propEq, sortBy } from "ramda";
import React from "react";
import { Flex } from "rebass";
import styled from "styled-components";

import Input from "../../elements/Input";
import stringFrIncludes from "../../libs/stringFrIncludes";
import Tag from "./Tag";

const Container = styled(Flex)`
  flex-grow: 1;
`;
const SelectedTags = styled(Flex)`
  flex-wrap: wrap;
`;
const SuggestedTags = styled(Flex)`
  background-color: var(--color-pearl);
  border: solid 1px var(--color-misty-moss);
  border-bottom: 0;
  margin-top: 2.25rem;
  max-width: 32rem;
  position: absolute;
  user-select: none;
  z-index: 1;
`;
const SuggestedTagsRow = styled.div`
  border-bottom: solid 1px var(--color-misty-moss);
  cursor: pointer;
  font-size: 0.875rem;
  overflow: hidden;
  padding: 0.25rem 0.5rem;
  text-overflow: ellipsis;
  white-space: nowrap;

  :hover {
    background-color: var(--color-shadow);
    color: white;
  }
`;

export default class Tags extends React.PureComponent {
  constructor(props) {
    super(props);

    const tags =
      props.tags !== undefined
        ? this.sortByValue(
            props.tags.filter(
              ({ value }) => !any(propEq("value", value))(props.selectedTags)
            )
          )
        : [];

    this.state = {
      inputKey: 0,
      isLoading: true,
      selectedTags:
        props.selectedTags !== undefined
          ? this.sortByValue(props.selectedTags)
          : [],
      suggestedTags: [],
      tags
    };

    this.ariaName =
      this.props.ariaName !== undefined ? this.props.ariaName : "l'étiquette";
  }

  getTagByValue(_value, source) {
    return source.filter(({ value }) => value === _value)[0];
  }

  sortByValue(source) {
    return sortBy(prop("value"))(source);
  }

  addTag(_value) {
    const tag = this.getTagByValue(_value, this.state.tags);

    this.setState({
      inputKey: this.state.inputKey + 1,
      selectedTags: this.sortByValue([...this.state.selectedTags, tag]),
      suggestedTags: [],
      tags: this.state.tags.filter(({ value }) => value !== _value)
    });

    this.props.onAdd(tag);
  }

  removeTag(value) {
    const foundTag = this.getTagByValue(value, this.state.selectedTags);
    const tag = foundTag !== undefined ? foundTag : { value };

    this.setState({
      selectedTags: this.state.selectedTags.filter(
        ({ value: _value }) => _value !== value
      ),
      tags: this.sortByValue([...this.state.tags, tag])
    });

    this.props.onRemove(tag);
  }

  refreshSuggestedTags() {
    if (this.state.tags.length === 0) return;

    if (this.$input.value.length === 0) {
      this.setState({ suggestedTags: [] });

      return;
    }

    this.setState({
      suggestedTags: this.state.tags
        .filter(({ value }) => stringFrIncludes(this.$input.value, value))
        .slice(0, 9)
    });
  }

  getSuggestedTags() {
    return this.state.suggestedTags.map(({ value }, index) => (
      <SuggestedTagsRow key={index} onClick={() => this.addTag(value)}>
        {value}
      </SuggestedTagsRow>
    ));
  }

  getSelectedTags() {
    const selectedTags = Boolean(this.props.isEditable)
      ? this.state.selectedTags
      : this.props.selectedTags;

    return selectedTags.map((tag, index) => (
      <Tag
        ariaName={this.ariaName}
        key={index}
        onRemove={this.removeTag.bind(this)}
        url={tag.url}
        value={tag.value}
      />
    ));
  }

  render() {
    return (
      <Container flexDirection="column">
        {Boolean(this.props.isEditable) && (
          <Input
            key={this.state.inputKey}
            onChange={this.refreshSuggestedTags.bind(this)}
            placeholder={`Commencez à taper le nom de ${this.ariaName}`.replace(
              "de le",
              "du"
            )}
            ref={node => (this.$input = node)}
          />
        )}
        {Boolean(this.props.isEditable) &&
          this.state.suggestedTags.length !== 0 && (
            <SuggestedTags flexDirection="column">
              {this.getSuggestedTags()}
            </SuggestedTags>
          )}
        {!Boolean(this.props.hideTags) && (
          <SelectedTags>{this.getSelectedTags()}</SelectedTags>
        )}
      </Container>
    );
  }
}
