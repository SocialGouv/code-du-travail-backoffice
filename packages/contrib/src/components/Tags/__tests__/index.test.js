import React from "react";
import { cleanup, fireEvent, render } from "@testing-library/react";

import "../../../../__mocks__/waitFor";

import Tags from "..";

describe("[Contrib] components/<Tags />", () => {
  const props = {
    ariaName: "le tag",
    isEditable: true,
    onAdd: jest.fn(),
    onRemove: jest.fn(),
    selectedTags: [{ value: "A Tag" }, { value: "Another Tag" }],
    tags: [
      { value: "A Tag" },
      { value: "Another Tag" },
      { value: "Yet Another Tag" }
    ]
  };

  const γ = render(<Tags {...props} />);

  it("should match snapshot", () => {
    expect(γ.container).toMatchSnapshot();
    expect(γ.queryByText(props.tags[0].value)).toBeInTheDocument();
    expect(γ.queryByText(props.tags[1].value)).toBeInTheDocument();
    expect(γ.queryByText(props.tags[2].value)).not.toBeInTheDocument();
  });

  it("should show the expected suggestion", () => {
    fireEvent.input(γ.getByPlaceholderText(`Commencez à taper le nom du tag`), {
      target: { value: "yet" }
    });

    expect(γ.queryByText(props.tags[2].value)).toBeInTheDocument();
  });

  it("should add the expected tag", async () => {
    fireEvent.click(γ.queryByText(props.tags[2].value));

    expect(props.onAdd).toHaveBeenCalledWith(props.tags[2]);
    expect(
      γ.getByPlaceholderText(`Commencez à taper le nom du tag`).value
    ).toBe("");
    expect(γ.queryByText(props.tags[0].value)).toBeInTheDocument();
    expect(γ.queryByText(props.tags[1].value)).toBeInTheDocument();
    expect(γ.queryByText(props.tags[2].value)).toBeInTheDocument();
    expect(
      γ.queryByAltText(
        `Bouton supprimant ${props.ariaName} ${props.tags[2].value}`
      )
    ).toBeInTheDocument();
  });

  it("should remove the expected tag", async () => {
    fireEvent.click(
      γ.getByAltText(
        `Bouton supprimant ${props.ariaName} ${props.tags[0].value}`
      )
    );

    expect(props.onRemove).toHaveBeenCalledWith(props.tags[0]);
    expect(γ.queryByText(props.tags[0].value)).not.toBeInTheDocument();
    expect(γ.queryByText(props.tags[1].value)).toBeInTheDocument();
    expect(γ.queryByText(props.tags[2].value)).toBeInTheDocument();
  });

  it("should show the expected suggestions", () => {
    fireEvent.input(γ.getByPlaceholderText(`Commencez à taper le nom du tag`), {
      target: { value: "a" }
    });

    expect(
      γ.getByPlaceholderText(`Commencez à taper le nom du tag`).value
    ).toBe("a");
    expect(γ.queryByText(props.tags[0].value)).toBeInTheDocument();
  });

  it("should hide the suggestion when the input is emptied", () => {
    fireEvent.input(γ.getByPlaceholderText(`Commencez à taper le nom du tag`), {
      target: { value: "" }
    });

    expect(
      γ.getByPlaceholderText(`Commencez à taper le nom du tag`).value
    ).toBe("");
    expect(γ.queryByText(props.tags[0].value)).not.toBeInTheDocument();
  });
});

describe("[Contrib] components/<Tags /> without optional props", () => {
  const props = {
    isEditable: true,
    onAdd: jest.fn(),
    onRemove: jest.fn()
  };

  let γ;

  it("should match snapshot", () => {
    cleanup();

    γ = render(<Tags {...props} />);

    expect(γ.container).toMatchSnapshot();
  });

  // it("should ot show any suggestion", () => {
  //   fireEvent.input(
  //     γ.getByPlaceholderText(`Commencez à taper le nom de l'étiquette`),
  //     {
  //       target: { value: "a" }
  //     }
  //   );
  // });
});

describe("[Contrib] components/<Tags /> without `isEditable`", () => {
  const props = {
    onAdd: jest.fn(),
    onRemove: jest.fn(),
    selectedTags: [{ value: "A Tag" }, { value: "Another Tag" }],
    tags: [
      { value: "A Tag" },
      { value: "Another Tag" },
      { value: "Yet Another Tag" }
    ]
  };

  let γ;

  it("should match snapshot", () => {
    cleanup();

    γ = render(<Tags {...props} />);

    expect(γ.container).toMatchSnapshot();
  });
});
