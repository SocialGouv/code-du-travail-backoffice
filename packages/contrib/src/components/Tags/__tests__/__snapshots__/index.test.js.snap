// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Contrib] components/<Tags /> should add the expected tag 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -35,8 +35,22 @@
            alt=\\"Bouton supprimant le tag Another Tag\\"
            class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
            src=\\"[object Object]\\"
          />
        </div>
+       <div
+         class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
+       >
+         <span
+           class=\\"Tag__Text-lc1n3o-1 dtXAeC\\"
+         >
+           Yet Another Tag
+         </span>
+         <img
+           alt=\\"Bouton supprimant le tag Yet Another Tag\\"
+           class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
+           src=\\"[object Object]\\"
+         />
+       </div>
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`[Contrib] components/<Tags /> should hide the suggestion when the input is emptied 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -13,28 +13,28 @@
          class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
        >
          <span
            class=\\"Tag__Text-lc1n3o-1 dtXAeC\\"
          >
-           A Tag
+           Another Tag
          </span>
          <img
-           alt=\\"Bouton supprimant le tag A Tag\\"
+           alt=\\"Bouton supprimant le tag Another Tag\\"
            class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
            src=\\"[object Object]\\"
          />
        </div>
        <div
          class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
        >
          <span
            class=\\"Tag__Text-lc1n3o-1 dtXAeC\\"
          >
-           Another Tag
+           Yet Another Tag
          </span>
          <img
-           alt=\\"Bouton supprimant le tag Another Tag\\"
+           alt=\\"Bouton supprimant le tag Yet Another Tag\\"
            class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
            src=\\"[object Object]\\"
          />
        </div>
      </div>"
`;

exports[`[Contrib] components/<Tags /> should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-bwzfXH Tags__Container-sc-9wg2t-0 kGdxZz"
  >
    <input
      class="Input___Input-l1y0rz-1 didSAw"
      placeholder="Commencez à taper le nom du tag"
    />
    <div
      class="sc-bdVaJa sc-bwzfXH Tags__SelectedTags-sc-9wg2t-1 gNdWxU"
    >
      <div
        class="sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb"
      >
        <span
          class="Tag__Text-lc1n3o-1 dtXAeC"
        >
          A Tag
        </span>
        <img
          alt="Bouton supprimant le tag A Tag"
          class="Tag__Button-lc1n3o-2 jbGhej"
          src="[object Object]"
        />
      </div>
      <div
        class="sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb"
      >
        <span
          class="Tag__Text-lc1n3o-1 dtXAeC"
        >
          Another Tag
        </span>
        <img
          alt="Bouton supprimant le tag Another Tag"
          class="Tag__Button-lc1n3o-2 jbGhej"
          src="[object Object]"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`[Contrib] components/<Tags /> should remove the expected tag 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -13,28 +13,28 @@
          class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
        >
          <span
            class=\\"Tag__Text-lc1n3o-1 dtXAeC\\"
          >
-           A Tag
+           Another Tag
          </span>
          <img
-           alt=\\"Bouton supprimant le tag A Tag\\"
+           alt=\\"Bouton supprimant le tag Another Tag\\"
            class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
            src=\\"[object Object]\\"
          />
        </div>
        <div
          class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
        >
          <span
            class=\\"Tag__Text-lc1n3o-1 dtXAeC\\"
          >
-           Another Tag
+           Yet Another Tag
          </span>
          <img
-           alt=\\"Bouton supprimant le tag Another Tag\\"
+           alt=\\"Bouton supprimant le tag Yet Another Tag\\"
            class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
            src=\\"[object Object]\\"
          />
        </div>
      </div>"
`;

exports[`[Contrib] components/<Tags /> should show the expected suggestion 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,10 +5,19 @@
      <input
        class=\\"Input___Input-l1y0rz-1 didSAw\\"
        placeholder=\\"Commencez à taper le nom du tag\\"
      />
      <div
+       class=\\"sc-bdVaJa sc-bwzfXH Tags__SuggestedTags-sc-9wg2t-2 bGEYen\\"
+     >
+       <div
+         class=\\"Tags__SuggestedTagsRow-sc-9wg2t-3 cIyIVc\\"
+       >
+         Yet Another Tag
+       </div>
+     </div>
+     <div
        class=\\"sc-bdVaJa sc-bwzfXH Tags__SelectedTags-sc-9wg2t-1 gNdWxU\\"
      >
        <div
          class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
        >"
`;

exports[`[Contrib] components/<Tags /> should show the expected suggestions 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,36 +5,45 @@
      <input
        class=\\"Input___Input-l1y0rz-1 didSAw\\"
        placeholder=\\"Commencez à taper le nom du tag\\"
      />
      <div
+       class=\\"sc-bdVaJa sc-bwzfXH Tags__SuggestedTags-sc-9wg2t-2 bGEYen\\"
+     >
+       <div
+         class=\\"Tags__SuggestedTagsRow-sc-9wg2t-3 cIyIVc\\"
+       >
+         A Tag
+       </div>
+     </div>
+     <div
        class=\\"sc-bdVaJa sc-bwzfXH Tags__SelectedTags-sc-9wg2t-1 gNdWxU\\"
      >
        <div
          class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
        >
          <span
            class=\\"Tag__Text-lc1n3o-1 dtXAeC\\"
          >
-           A Tag
+           Another Tag
          </span>
          <img
-           alt=\\"Bouton supprimant le tag A Tag\\"
+           alt=\\"Bouton supprimant le tag Another Tag\\"
            class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
            src=\\"[object Object]\\"
          />
        </div>
        <div
          class=\\"sc-bdVaJa sc-bwzfXH Tag__Container-lc1n3o-0 dVwbtb\\"
        >
          <span
            class=\\"Tag__Text-lc1n3o-1 dtXAeC\\"
          >
-           Another Tag
+           Yet Another Tag
          </span>
          <img
-           alt=\\"Bouton supprimant le tag Another Tag\\"
+           alt=\\"Bouton supprimant le tag Yet Another Tag\\"
            class=\\"Tag__Button-lc1n3o-2 jbGhej\\"
            src=\\"[object Object]\\"
          />
        </div>
      </div>"
`;

exports[`[Contrib] components/<Tags /> without optional props should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-bwzfXH Tags__Container-sc-9wg2t-0 kGdxZz"
  >
    <input
      class="Input___Input-l1y0rz-1 didSAw"
      placeholder="Commencez à taper le nom de l'étiquette"
    />
    <div
      class="sc-bdVaJa sc-bwzfXH Tags__SelectedTags-sc-9wg2t-1 gNdWxU"
    />
  </div>
</div>
`;

exports[`[Contrib] components/<Tags /> without optional props should ot show any suggestion 1`] = `
"Snapshot Diff:
  Comparing two different types of values. Expected object but received null."
`;
